%!TEX root = Main_Assignment4.tex
\documentclass[Main_Assignment4]{subfiles}

\begin{document}

\subsection{Part c: Unit test case}
When writing unit test cases it's important not to strive towards verifying the code works, but test \emph{if} it can fail.

Example for this is applying boundary test for the UUT/CUT (Unit Under Test/Code Under Test). 

\lstinputlisting[caption=Unit tests -- different k-values, style=Code-C++, label=lst:UUT1]{testClass.cpp}


Beside \codeTitle \ref{lst:UUT1} different scenarios can be constructed, some more plausible than others:
\begin{itemize}
	\item \code{x}  is a \code{char} (8-bit) and $x \cdot k$ exceeds 256. This will cause an overflow on \code{x} and return a wrong value to the assertion.
	\item[] Example: \code{assert 300 = 2 + (2 * 149)}.
	\item[] Even though m actually has the correct value the assertion will be triggered.

	\item \code{k} is greater than the \code{int}, which causes an overflow in \code{k} and will add \code{x} the wrong number of times.

	\item Both \code{k} and \code{x} are global variables and can, in a threaded system, be changed when the function is running, giving a wrong result.
\end{itemize}

\subsection{Part d}
Refer to Problem 1.

\end{document}